src/xSTIR/cSTIR/stir_data_containers.h

```diff
@@ -465,64 +483,188 @@ namespace sirf {
 
-	class PETImageData : public aDataContainer < float > {
+	typedef Image3DF::full_iterator Image3DFIterator;
+	typedef Image3DF::const_full_iterator Image3DFIterator_const;
+
+	//class STIRImageData : public aDataContainer < float > {
+	class STIRImageData : public PETImageData { //<Iterator, Iterator_const> {
 	public:
-		PETImageData(){}
-		PETImageData(const PETImageData& image)
+		//typedef PETImageData<Iterator, Iterator_const>::Iter BaseIter;
+		//typedef PETImageData<Iterator, Iterator_const>::Iter_const BaseIter_const;
+		typedef ImageData::Iterator BaseIter;
+		typedef ImageData::Iterator_const BaseIter_const;
+		class Iterator : public BaseIter {
+		public:
+			Iterator(const Image3DFIterator& iter) : _iter(iter)
+			{}
+			Iterator& operator=(const Iterator& iter)
+			{
+				_iter = iter._iter;
+				_ref.copy(iter._ref);
+				_sptr_iter = iter._sptr_iter;
+				return *this;
+			}
+			virtual Iterator& operator++()
+			{
+				++_iter;
+				return *this;
+			}
+			//virtual Iterator& operator++(int)
+			//{
+			//	_sptr_iter.reset(new Iterator(_iter));
+			//	++_iter;
+			//	return *_sptr_iter;
+			//}
+			virtual bool operator==(const BaseIter& an_iter) const
+			{
+				const Iterator& iter = (const Iterator&)an_iter;
+				return _iter == iter._iter;
+			}
+			virtual bool operator!=(const BaseIter& an_iter) const
+			{
+				const Iterator& iter = (const Iterator&)an_iter;
+				return _iter != iter._iter;
+			}
+			virtual FloatRef& operator*()
+			{
+				float& v = *_iter;
+				_ref.set_ptr(&v);
+				return _ref;
+			}
+		private:
+			Image3DFIterator _iter;
+			FloatRef _ref;
+			std::shared_ptr<Iterator> _sptr_iter;
+		};
+		class Iterator_const : public BaseIter_const {
+		public:
+			Iterator_const(const Image3DFIterator_const& iter) : _iter(iter)
+			{}
+			Iterator_const& operator=(const Iterator_const& iter)
+			{
+				_iter = iter._iter;
+				_ref.copy(iter._ref);
+				_sptr_iter = iter._sptr_iter;
+				return *this;
+			}
+			virtual Iterator_const& operator++()
+			{
+				++_iter;
+				return *this;
+			}
+			//virtual Iterator_const& operator++(int)
+			//{
+			//	_sptr_iter.reset(new Iterator_const(_iter));
+			//	++_iter;
+			//	return *_sptr_iter;
+			//}
+			virtual bool operator==(const BaseIter_const& an_iter) const
+			{
+				const Iterator_const& iter = (const Iterator_const&)an_iter;
+				return _iter == iter._iter;
+			}
+			virtual bool operator!=(const BaseIter_const& an_iter) const
+			{
+				const Iterator_const& iter = (const Iterator_const&)an_iter;
+				return _iter != iter._iter;
+			}
+			virtual const FloatRef& operator*() const
+			{
+				const float& v = *_iter;
+				_ref.set_ptr((void*)&v);
+				return _ref;
+			}
+		private:
+			Image3DFIterator_const _iter;
+			mutable FloatRef _ref;
+			std::shared_ptr<Iterator_const> _sptr_iter;
+		};
+		STIRImageData() {}
+		STIRImageData(const ImageData& id)
+		{
+			Dimensions dim = id.dimensions();
+			int nx = dim["x"];
+			int ny = dim["y"];
+			int nz = 1;
```

Comment:
I guess you did this because we don't necessarily know what the 3rd dimension of the MR data is?

So, if I had an image that had `z=100` for three time points, I would create a `STIRImageData` with 300 slices? Is this desired? 

Could we throw an error if `ndim > 3`?

Reply:
images reconstructed by gadgetron normally have 4 dimensions: nx, ny, nz (never saw nz > 1) and number of slices

note also that i actually added that piece of code merely to test conversion (constructing from an abstract image using abstract iterators) - dimensions and other geometrical metadata will have to be handled thoroughly (i believe Ashley is working on it on his branch)